### ISSUE TITLE
Bitwise operation on pixels

### ISSUE THREAD

**Issue by Hagb** [nwrkbiz/android-xserver#36]
Refer to https://www.x.org/releases/current/doc/xproto/x11protocol.html#requests:CreateGC .
With `function` and `plane-mask` on `CreateGC`, client can tell server to draw with bitwise operations:
```
	((src FUNC dst) AND plane-mask) OR (dst AND (NOT plane-mask))
```

| Function | Operation |
|:----:|:----:|
| Clear  | 0 |
| And  | src AND dst |
| AndReverse  | src AND (NOT dst) |
| Copy  | src |
| AndInverted  | (NOT src) AND dst |
| NoOp  | dst |
| Xor  | src XOR dst |
| Or  | src OR dst |
| Nor  | (NOT src) AND (NOT dst) |
| Equiv  | (NOT src) XOR dst |
| Invert  | NOT dst |
| OrReverse  | src OR (NOT dst) |
| CopyInverted  | NOT src |
| OrInverted  | (NOT src) OR dst |
| Nand  | (NOT src) OR (NOT dst) |
| Set  | 1 |

I failed to find an efficient way to implement these bitwise operation after spending many hours. I think it might be difficult to implement a paint with bitwise Xfermode. As a result, it is hard to draw via methods of canvas.

TL;DR:

Android doesn't have bitwise [`Xfermode`](https://developer.android.com/reference/android/graphics/Xfermode) now. `PixelXorXfermode` has been removed, and `Xfermode` has only one subclass [`PorterDuffXfermode`](https://developer.android.com/reference/android/graphics/PorterDuffXfermode) now, which has no bitwise operation, and [`PorterDuff.Mode.XOR`](https://developer.android.com/reference/android/graphics/PorterDuff.Mode#XOR) is xor operation for area but not for color, which is misused in

https://github.com/nwrkbiz/android-xserver/blob/51e907880d4295ee3c381a593c738ca365c066fb/library/src/main/java/au/com/darkside/xserver/GContext.java#L406

I tried to write our own `Xfermode` subclass, so that we can still use methods of canvas to draw. But finally found that [`Paint`](https://developer.android.com/reference/android/graphics/Paint) [uses native code](https://cs.android.com/android/platform/superproject/+/master:frameworks/base/graphics/java/android/graphics/Paint.java;drc=76ba45e4bd7ca9a1ae8f216cd348ccf53e961af4;bpv=1;bpt=1;l=1361?gsn=installXfermode&gs=kythe%3A%2F%2Fandroid.googlesource.com%2Fplatform%2Fsuperproject%3Flang%3Djava%3Fpath%3Dandroid.graphics.Paint%23ae155468af1ebc32502092cb2de225e803deb5cc8c361fafacc319151c4837b3) and [hardcodes `PorterDuffXfermode`](https://cs.android.com/android/platform/superproject/+/master:frameworks/base/libs/hwui/jni/Paint.cpp;bpv=1;bpt=1;l=799?q=setXfermode&gsn=setXfermode&gs=kythe%3A%2F%2Fandroid.googlesource.com%2Fplatform%2Fsuperproject%3Flang%3Dc%252B%252B%3Fpath%3Dframeworks%2Fbase%2Flibs%2Fhwui%2Fjni%2FPaint.cpp%23zZVBxTkYoV0k92Wet7cHM_v3E6NtyYQwefHdi8E4l6k), even if `Paint.setXfermode` accepts `Xfermode`.

[`ColorMatrixColorFilter`](https://developer.android.com/reference/android/graphics/ColorMatrixColorFilter) can be used to implement bitwise `NOT` on `src` but not `dst`.

**Comment by nwrkbiz** [comment]
Hmm I think paint.setColorFilter() with an appropiate filter matrix might provide the same functionality like bitwise XOR:

```java
private ColorMatrix getColorMatrix() {
    return new ColorMatrix(new float[] {
        -1,  0,  0,  0, 255,
        0, -1,  0,  0, 255,
        0,  0, -1,  0, 255,
        0,  0,  0,  1,   0
    });
}

// ...
// ...

_paint.setColorFilter(new ColorMatrixColorFilter(getColorMatrix()));
```

**Comment by Hagb** [comment]
> Hmm I think paint.setColorFilter() with an appropiate filter matrix might provide the same functionality like bitwise XOR:
> 
> ```java
> private ColorMatrix getColorMatrix() {
>     return new ColorMatrix(new float[] {
>         -1,  0,  0,  0, 255,
>         0, -1,  0,  0, 255,
>         0,  0, -1,  0, 255,
>         0,  0,  0,  1,   0
>     });
> }
> 
> // ...
> // ...
> 
> _paint.setColorFilter(new ColorMatrixColorFilter(getColorMatrix()));
> ```

if I was not mistaken, color filters is only applied to source but not destination, and the above matrix is bitwise `NOT`. (And any linear matrix operations can not simulate `XOR`, `AND` or `OR`.) 
When plane-mask is 0xffffff, with `ColorMatrix`, we can implement CopyInverted (NOT src), but not some other functions including`XOR`, `AND` and `OR`.

For example, with the above `_paint`, `_canvas.drawLines(points, paint)` means drawing a line with CopyInverted (NOT src).
